### PR TITLE
command/format: improve list nested attribute rendering

### PR DESCRIPTION
**Pull request by bluekeyes**
This change contains two commits, the first addressing a bug and the second improving general rendering in the same area. I'm submitting them together, but would be happy to split them up or only submit the bug fix.

### Bug: Lists that mix changed and unchanged elements did not render

Diffs only rendered correctly if all changed elements appeared before all unchanged elements. Once an unchanged element was found, all remaining elements were skipped. This usually led to the output being an empty list with a weird amount of space between the brackets.

**Before:**

```
  # nested_fields.example will be updated in-place
  ~ resource "nested_fields" "example" {
        name   = "example"
      ~ values = [
          ~ {          },
          # (3 unchanged elements hidden)
        ]
    }
```

**After:**

```
  # nested_fields.example will be updated in-place
  ~ resource "nested_fields" "example" {
        name   = "example"
      ~ values = [
          ~ {
              name  = "b"
            ~ value = "2" -> "3"
          },
          # (3 unchanged elements hidden)
        ]
    }
```

### Formatting Improvements

I made several changes to make list diffs clearer and more consistent:

* Separate items with brackets instead of only new lines. This better matches the input syntax and avoids confusion from the first and last brackets implying there is a single item.
* Render an action symbol for each element of the list
* Use the correct action symbol for new and deleted items
* Fix the alignment of opening and closing brackets

**Before:**

```
  # nested_fields.example will be updated in-place
  ~ resource "nested_fields" "example" {
        name   = "example"
      ~ values = [
          ~ {
              name  = "c"
            ~ value = "3" -> "c"

            + name  = "e"
            + value = "5"
          },
          # (3 unchanged elements hidden)
        ]
    }
```

**After:**

```
  # nested_fields.example will be updated in-place
  ~ resource "nested_fields" "example" {
        name   = "example"
      ~ values = [
          ~ {
                name  = "c"
              ~ value = "3" -> "c"
            },
          + {
              + name  = "e"
              + value = "5"
            },
            # (3 unchanged elements hidden)
        ]
    }
```

### Context

I'm writing a provider using the new `terraform-plugin-framework` that uses the [`ListNestedAttributes`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v0.4.2/tfsdk#ListNestedAttributes) type in several places. I extracted the relevant logic into a [sample provider](https://gist.github.com/bluekeyes/44d54f191a1628ed2e732ef0f2324f55) to experiment more after I observed strange plan output.

### CODE REVIEW

File: internal/command/format/diff.go
Comment (alisdair) [inline internal/command/format/diff.go:537]: We're reusing the same `blockBodyDiffResult ` value for multiple calls here, which means that the `skippedAttributes` field will continue to increment with every element. Screenshot example, where all of these values have one skipped attribute:

image

**Review by alisdair — CHANGES_REQUESTED** [review]
Thanks very much for making these fixes and improvements! I spotted one struct reuse bug to fix before merge.

One problem that this PR is that we're not showing any context for changed nested list items, which we do when rendering sequence attributes. The code that handles that for lists and tuples is here: https://github.com/hashicorp/terraform/blob/834f61cd54dfa196511410f843f24b532e5b4ecc/internal/command/format/diff.go#L1354-L1436

If you happen to have ideas around how to bring that same functionality across here, it would be great to hear them. Doesn't block merging this PR, though!

File: internal/command/format/diff.go
Comment (bluekeyes) [reply internal/command/format/diff.go:539]: Good catch, fixed and extended the tests to cover this.

**Comment by bluekeyes** [comment]
> One problem that this PR is that we're not showing any context for changed nested list items, which we do when rendering sequence attributes.

I noticed this as well. An initial version of this change partially addressed it by inserting the `(N unchanged elements hidden)` message at each each place in the list where elements were suppressed, which at least lets the user infer the index that is changing. I removed it because it felt noisy and didn't seem to fit with the other suppression messages (which always appear at the end of the element), but it was pretty easy to implement if you'd like me to add it again.

Beyond that, I'll think about how to provide better context using the linked code as an example, but I don't know if I'll have time to implement it in the near future.

**Review by alisdair — APPROVED** [review]
Thanks, looks great! Merging. The additional sequence context we discussed is just a nice-to-have—maybe we'll find a way to add that in a future release.

**Pull request merged**
